### ISSUE TITLE
CircleCI Recommendations

### ISSUE THREAD

**Issue by Siphonophora** [spekt/junit.testlogger#37]
We are looking for help adding platform specific recommendations (if any), screenshots, and examples for CircleCI which are similar to those here. https://github.com/spekt/junit.testlogger/blob/master/docs/gitlab-recommendation.md

**Comment by Peter-Darton-i2** [comment]
I'm a CircleCI user who's also using `vstest.console` and in need of a means of getting JUnit reports.

I don't have a full picture of CircleCI's functionality, but I do know that CircleCI will report any `<system-out>` data that's included _inside_ a `<testcase>` ... and that it ignores any `<system-out>` data that's _outside_ testcases.
I can see that the data is present in the .trx file but in the JUnit xml, all the individual tests' system-out has been appended together in a per-testsuite element rather than a per-testcase element as CircleCI expects.
CircleCI will also report information provided within a `<failure>` element too.

... but CircleCI doesn't let you inspect any test results that pass (at least, not until the next day), so there's not much point in going into any detail for those, which means that the existing `FailureBodyFormat=Verbose` setting may be sufficient without further work to put the `<system-out>` data in a per-testcase location.

e.g. this is a (heavily redacted, tweaked for brevity) JUnit XML file generated by one of my CircleCI builds (but _not_ generated by this logger - this is a "what can CircleCI do" example, not a "what can this logger do" demo, sorry...):
```
<testsuite name="OPA policies" tests="25" failures="1" skipped="0" assertions="25" time="0" timestamp="2023-07-14T17:25:53+00:00" file="policies">
  <testcase name="public_repo_REDACTEDREDACTEDREDACTEDpublic__is_banned_from_using_contexts" classname="policies.test_public_repos_are_not_permitted_to_use_any_contexts" assertions="1" time="0">
    <system-out>
      <![CDATA[ [ { "decision": { "enabled_rules": [ "bad_context_use" ], "status": "PASS" }, "evaluation": { "meta": { "build_number": 123, "project_id": "11111111-2222-3333-4444-555555555555", "ssh_rerun": false, "vcs": { "branch": "main", "origi ]]>
      <![CDATA[ n_repository_url": "https://github.com/REDACTEDREDACTEDREDACTEDREDACT/", "target_repository_url": "https://github.com/REDACTEDREDACTEDREDACTEDREDACT/" } }, "org": { "bad_context_use": {}, "contexts_that_are_restricted_to_only_specified_projects": { REDACTED }, "enable_rule": [ "bad_context_use" ], "git_url_patterns_that_may_not_use_contexts": [ REDACTED ], "hard_fail": [ "bad_context_use", "project_may_not_use_contexts" ], "policy_name": [ "not_permitted_to_use_any_contexts", "secret_context_usage" ], "project_ids_that_may_not_use_contexts": [ REDACTED ], "project_may_not_use_contexts": { "my-workflow:job1:any-context-at-all": "REDACTEDrohibited the use of contexts in this project (found context any-context-at-all in my-workflow/job1)" } } }, "input": { "jobs": { "job1": { "docker": [ { "image": "cimg/base:current" } ], "steps": [ { "run": { "command": "echo \"Hello\"" } } ] } }, "workflows": { "my-workflow": { "jobs": [ { "job1": { "context": [ "any-context-at-all" ] } } ] } } } } ] ]]>
    </system-out>
    <failure message="Passed is false" type="AssertionError"> --- Expected +++ Actual @@ -2,7 +2,3 @@ - bad_context_use - - project_may_not_use_contexts -hard_failures: - - reason: REDACTEDrohibited the use of contexts in this project (found context any-context-at-all in my-workflow/job1) - rule: project_may_not_use_contexts -status: HARD_FAIL +status: PASS </failure>
    <properties>
      <property name="Passed" value="false"/>
      <property name="Group" value="policies"/>
      <property name="Name" value="test_public_repos_are_not_permitted_to_use_any_contexts/public_repo_REDACTEDREDACTEDREDACTEDpublic__is_banned_from_using_contexts"/>
      <property name="Elapsed" value="581.238µs"/>
    </properties>
  </testcase>
</testsuite>
```
...and it results in a CircleCI report that looks like this (blue stuff is where I've redacted stuff):
![image](https://github.com/spekt/junit.testlogger/assets/97306111/11c91fce-0c43-4e56-99d4-7bd67482fc8b)

...which can be expanded to to show the `<system-out>` data etc so it'll look like this:
![image](https://github.com/spekt/junit.testlogger/assets/97306111/28030950-5933-4f14-b6ae-97b3db176d59)

... with the full (redacted) text being:
```
[
{
  "decision": {
    "enabled_rules": [
      "bad_context_use"
    ],
    "status": "PASS"
  },
  "evaluation": {
    "meta": {
      "build_number": 123,
      "project_id": "11111111-2222-3333-4444-555555555555",
      "ssh_rerun": false,
      "vcs": {
        "branch": "main",
        "origin_repository_url": "https://github.com/REDACTEDREDACTED/",
        "target_repository_url": "https://github.com/REDACTEDREDACTED/"
      }
    },
    "org": {
      "bad_context_use": {},
      "contexts_that_are_restricted_to_only_specified_projects": {
        REDACTED
      },
      "enable_rule": [
        "bad_context_use"
      ],
      "git_url_patterns_that_may_not_use_contexts": [
        REDACTED
      ],
      "hard_fail": [
        "bad_context_use",
        "project_may_not_use_contexts"
      ],
      "policy_name": [
        "not_permitted_to_use_any_contexts",
        "secret_context_usage"
      ],
      "project_ids_that_may_not_use_contexts": [
        REDACTED
      ],
      "project_may_not_use_contexts": {
        "my-workflow:job1:any-context-at-all": "REDACTED prohibited the use of contexts in this project (found context any-context-at-all in my-workflow/job1)"
      }
    }
  },
  "input": {
    "jobs": {
      "job1": {
        "docker": [
          {
            "image": "cimg/base:current"
          }
        ],
        "steps": [
          {
            "run": {
              "command": "echo \"Hello\""
            }
          }
        ]
      }
    },
    "workflows": {
      "my-workflow": {
        "jobs": [
          {
            "job1": {
              "context": [
                "any-context-at-all"
              ]
            }
          }
        ]
      }
    }
  }
}
]Passed is false
--- Expected
+++ Actual
@@ -2,7 +2,3 @@
     - bad_context_use
-    - project_may_not_use_contexts
-hard_failures:
-    - reason: REDACTED prohibited the use of contexts in this project (found context any-context-at-all in my-workflow/job1)
-      rule: project_may_not_use_contexts
-status: HARD_FAIL
+status: PASS
```
...which seems to include both the `<system-out>` text, the `<failure>` message and its text, but none of the `<properties>` in the XML.

TL;DR: `FailureBodyFormat=Verbose` is likely to help CircleCI users.

**Comment by codito** [comment]
Available at https://github.com/spekt/junit.testlogger/blob/master/docs/circleci-recommendation.md.

Thanks @Peter-Darton-i2 for contributing the guide.
